### ISSUE TITLE
Add "update-order" to "docker node --availability drain"

### ISSUE THREAD

**Issue by markvr** [moby/moby#34139]
In the same way that `docker service update` has an `update-order` option, it would be really helpful if `docker node --availability drain` had the same.  This way, when draining a node for maintenance, services with only one replica would be moved with zero downtime, and services that required multiple replicas wouldn't lose capacity  during the migration.

It's possible to work around this by manually scaling each service on the node up before draining the node, but doing this repeatedly across every node is very time consuming.

Having it on the `node` command itself would be a huge help.

**Comment by markvr** [comment]
Another untested workaround might be: 
Run each service with a "--constraint 'node.labels.type != draining'" and set this label on a host that is about to be drained.  Then run "service update --update-order start-first" for each service on that node before finally setting the node availability to "drain".

If it works as expected, this would be simpler than scaling up and down the replicas for each service.

**Comment by markvr** [comment]
It turns out that adding a label to a node instantly affects all services on that node, so the workaround I suggest is no different to simply saying "--availability=drain". 

That's annoying, it means the only way to remove a node with zero downtime is:

- Manually scale each service on the node up by 1 and wait for them to deploy.  Keep a record of what service you altered, and what the original values were.
- Set `availability=drain` on the node.
- This will start more services on other nodes as they leave the 'draining' node.  Wait for these services to settle.
- Scale all the services back down again to their original levels.

If you are patching each node in turn, you need to repeat this for all nodes.  Obviously it could be scripted, but what a pain!!

**Comment by goetas** [comment]
I was rally expecting that this feature worked out of the box.

Since docker is able to update services with zero-downtime, was expecting that draining nodes had the same behavior.

This is really weird

**Comment by IgorOhrimenko** [comment]
If the image is small, for example, memcache, then it quickly dies out and quickly starts on another node, but if the image is large and it is not on the target node, then it is loaded first, which may take a long time or not at all, if with the registry or image something happened. As a result, the service downtime occurs, which should be avoided in every possible way.
Also, some services cannot be duplicated, so the option with scale may not suit you.
I suggest using this solution:
```
docker node update --availability pause $(hostname) > /dev/null && \
docker node ps $ (hostname) --filter desired-state=running --format {{.Name}} | \
awk -F. '{print $ 1}' | \
xargs -L1 docker service update --force --detach
```
$(hostname) > /dev/null so that the line with the node name does not get into the service update
awk -F. select the name of the service, not the tasks
xargs -L1, one argument per line, since "docker service update" requires exactly 1 argument.
--detach send commands without displaying information about the process, without it it will be a sequential launch of updates with information about raising a new instance of the service and waiting for it to work

**Comment by AlexHill** [comment]
Yeah, this is pretty painful. Draining should respect the update_config, or have an equivalent option.

The `pause`/`update --force` method is pretty good, but it results in a lot of churn because *all* the tasks for each service running on the departing node get rescheduled, not just the tasks that were running on that node. So if you have a lot of services with a lot of replicas, this can result in a lot of churn on your other nodes.

**Comment by r4nc0r** [comment]
Yes this would definitely improve working with a docker swarm!

**Comment by AlexHill** [comment]
@thaJeztah just a general question - are issues like this generally amenable to being addressed by PR from a member of the community? A fix would raise questions of backwards compatibility that would need to be answered, potentially add new modes to CLI commands, updating docs, probably require some design input, etc etc. I've never contributed to Docker and Swarm issues seem to spend a fair amount of time in limbo. No judgment here whatsoever, I'd like to work on this but I just don't have much of a feel for the Docker community yet so I'm not sure how much friction a motivated but green community contributor would face.

**Comment by thaJeztah** [comment]
Yes, contributions are welcome.

Haven't looked closely into backward compatibility myself, but if you think this would be a breaking change, then input/suggestions are welcome as well.

If changes are needed in Swarmkit (https://github.com/docker/swarmkit), that code is currently (primarily) maintained by Mirantis, but you can ping @dperny on pull requests (happy to have a look as well, although I'm not officially a maintainer on that project)

**Comment by AlexHill** [comment]
Thanks! I'll write up some kind of proposal in this thread soon. I expect the bulk of the changes will be in swarmkit, yep.

**Comment by svdo** [comment]
I also ran into this today. I'm curious to know if there is any progress/news on this. Thanks!

**Comment by everyx** [comment]
Any updates on this issue? k8s is too heavy and complex for small deployments, hope docker swarm mode can be improved about this.

**Comment by s4ke** [comment]
For anyone stumblinc across this - for our automation around upgrading nodes we are working on a simple script that waits for the swarm to be "settled". The current state lives in this comment: https://github.com/neuroforgede/swarmsible/issues/8#issuecomment-1454728644

**Comment by s4ke** [comment]
I have read the thread again and it turns out, I misunderstood the purpose of this ticket and it seems like https://github.com/moby/moby/issues/34139#issuecomment-318685179 is the only way currently to achieve zero downtime draining.

**Comment by s4ke** [comment]
As an alternative approach to move off services of nodes that are about to be drained it would be worth trying out to update services with "--constraint-add 'node.hostname!=$(hostname)'" or any other constraint on a per need basis instead of deploying them with the constraint from the get go. I haven't tried this on a multinode swarm yet, but it on a local "1 node swarm" it seems to be worth exploring more

This could work:

1. label your node with "draining=yes"
2. `docker node update $(hostname) --label-add draining=yes`
3. For each service run: `docker service update --constraint-add "node.labels.draining!=yes" <service_name>`
4. actually drain the node
5. `docker node update $(hostname) --label-rm draining`
6. `docker service update --constraint-add "node.labels.draining!=yes" <service_name>`
7. remove the label from your node

To not force the tasks off the nodes immediately, do this for every node in this order

**Comment by IgorOhrimenko** [comment]
The method 'constraint' immediately stop container, not start-first, when the image is weight or the entrypoint has long start, the service has downtime.

**Comment by s4ke** [comment]
In my experiment on a single node yesterday, it didn't stop the replicas on the new node immediately, instead it was "stuck" while waiting for a new node to be available for scheduling (which obviously didnt happen on a single node).

Will try the approach later today on a real cluster. @IgorOhrimenko the approach I describe is different to the one described in https://github.com/moby/moby/issues/34139#issuecomment-315738951 as it changes the order of node application and adding the constraint to the service.

**Comment by IgorOhrimenko** [comment]
The method 'constrain' work well for the global service, but not replicas.

**Comment by s4ke** [comment]
Have you tried the changed approach I described or are you assuming it is the same as https://github.com/moby/moby/issues/34139#issuecomment-315738951, because it isn't.
